fix(tests): set $SHELL = sh in tests

faergeek opened the pull request:
Trying to run `make test` under fish shell results in a failure with the following output:

```
FAILED   test/functional/terminal/tui_spec.lua @ 1017: TUI paste: terminal mode
test/functional/terminal/tui_spec.lua:1024: Row 1 did not match.
Expected:
  |*tty ready                                         |
  |*{1: }                                                 |
  |*                                                  |
  |                                                  |
  |{5:^^^^^^^                                           }|
  |{3:-- TERMINAL --}                                    |
  |{3:-- TERMINAL --}                                    |
Actual:
  |*process does not own the terminal                 |
  |*                                                  |
  |*[Process exited 2]{1: }                               |
  |                                                  |
  |{5:^^^^^^^                                           }|
  |{3:-- TERMINAL --}                                    |
  |{3:-- TERMINAL --}                                    |

To print the expect() call that would assert the current screen state, use
screen:snapshot_util(). In case of non-deterministic failures, use
screen:redraw_debug() to show all intermediate screen states.

stack traceback:
	test/functional/ui/screen.lua:622: in function '_wait'
	test/functional/ui/screen.lua:352: in function 'expect'
	test/functional/terminal/tui_spec.lua:1024: in function <test/functional/terminal/tui_spec.lua:1017>

FAILED   test/functional/terminal/tui_spec.lua @ 1551: TUI forwards :term palette colors with termguicolors
test/functional/terminal/tui_spec.lua:1567: Row 1 did not match.
Expected:
  |*{1:t}ty ready                                         |
  |                                                  |
  |*                                                  |
  |                                                  |
  |{2:^^^^^^^                                           }|
  |                                                  |
  |{3:-- TERMINAL --}                                    |
Actual:
  |*{1:p}rocess does not own the terminal                 |
  |                                                  |
  |*[Process exited 2]                                |
  |                                                  |
  |{2:^^^^^^^                                           }|
  |                                                  |
  |{3:-- TERMINAL --}                                    |

To print the expect() call that would assert the current screen state, use
screen:snapshot_util(). In case of non-deterministic failures, use
screen:redraw_debug() to show all intermediate screen states.

stack traceback:
	test/functional/ui/screen.lua:622: in function '_wait'
	test/functional/ui/screen.lua:352: in function 'expect'
	test/functional/terminal/tui_spec.lua:1567: in function <test/functional/terminal/tui_spec.lua:1551>
```

Which may be kind of related to https://github.com/neovim/neovim/issues/2354

I suppose that fish does not make the process it starts a session leader, based on this [comment](https://github.com/neovim/neovim/issues/2354#issuecomment-92386195). I checked it by running `:terminal ./build/bin/tty-test` from neovim with `shell=/bin/fish` (inherited from `$SHELL`) and it indeed complains about "process does not own the terminal". With `shell=sh` it doesn't complain. And unsetting `$SHELL` seems to make `neovim` to fall back to `shell=sh`.

I have also found another issue about test failures under fish, though the reason seem to be quite different https://github.com/neovim/neovim/issues/6172, but there is an associated PR https://github.com/neovim/neovim/pull/6176, in which in most recent review comments @justinmk suggests unsetting `$SHELL` in a `clear()` helper, instead of doing `set shell=sh` (which broke tests on windows at the time as I understand), which I did here. Though I'm not sure if exactly this is what was meant in that comment. Should it be done through `env` in `spawn_argv` or in `new_argv` instead?

I'm also not sure if two of the test runs fail because of this change or this tests just might be flaky. Failures occur in different tests in each of them.

zeertzjq commented:
`test/functional/vimscript/system_spec.lua` is failing on macOS CI. That test needs change.

justinmk commented on test/functional/helpers.lua:
```diff
@@ -447,6 +447,7 @@ end
 --    clear('-e')
 --    clear{args={'-e'}, args_rm={'-i'}, env={TERM=term}}
 function module.clear(...)
+  luv.os_unsetenv('SHELL')
```
Not a blocker, but would it be more correct to unset SHELL in `cmake/RunTests.cmake` ?

faergeek replied:
> It's probably less confusing to set `SHELL=sh` explicitly? (Except on Windows)

Would nvim even care about $SHELL value on Windows? `:help 'shell'` says it's $SHELL or sh, except on Windows where it's "cmd.exe" by default. And I thought that initial intention was to make nvim use the fallback value (only omitting $SHELL), not explicitly setting it to that value. May be adding a comment would help to make it less confusing?

> Not a blocker, but would it be more appropriate to make the change in `cmake/RunTests.cmake` ?

I'll try it.

faergeek replied:
I `grep`d in test/oldtest, here are a few interesting things I've found:

$SHELL is set to sh in Makefile:
https://github.com/neovim/neovim/blob/4258f4d53cb3d8d0adbaccf410e1bb422db9badb/test/old/testdir/Makefile#L12
I don't see anything special here regarding $SHELL for Windows (and it shouldn't matter on Windows anyway I guess).

Though in setup.vim $SHELL is reset to an empty string for win32:
https://github.com/neovim/neovim/blob/4258f4d53cb3d8d0adbaccf410e1bb422db9badb/test/old/testdir/setup.vim#L70-L77

Also `shell` option is set to 'sh' directly in unix.vim:
https://github.com/neovim/neovim/blob/4258f4d53cb3d8d0adbaccf410e1bb422db9badb/test/old/testdir/unix.vim#L2-L3

So for me it still seems safer to unset it or set to an empty string.

justinmk replied:
> I don't see anything special here regarding $SHELL for Windows (and it shouldn't matter on Windows anyway I guess).

The point is that explicitly setting SHELL to nonsense on Windows is unnecessary potential confusion so why not avoid it there? 



> So for me it still seems safer to unset it or set to an empty string.

Why? Eventhing in test/old/ is a separate set of tests. If SHELL=sh is done there, that implies setting SHELL=sh in RunTest.cmake may fix the issue, and matches the approach already taken in test/old/.

faergeek replied:
>But then if clear{env={SHELL=foo}} is specified by a caller, that should be respected.

I don't see why it wouldn't be respected right now. I moved unsetting `$SHELL` to `cmake/RunTests.cmake` as you suggested. So parent process won't inherit `$SHELL` during test run, but if value for `$SHELL` is explicitly passed to `clear` it should just work. Or am I missing something?

justinmk replied:
> > But then if clear{env={SHELL=foo}} is specified by a caller, that should be respected.
> 
> I don't see why it wouldn't be respected right now.

there i was referring to the case where SHELL is set in clear(). not relevant if SHELL is set in RunTests.cmake

zeertzjq commented on cmake/RunTests.cmake:
```diff
@@ -64,6 +64,14 @@ endif()
+#
+# Unsetting $SHELL makes neovim fall back to default value for 'shell' option.
+# See `:help 'shell'`.
+set(ENV{SHELL})
```
```suggestion
unset(ENV{SHELL})
```

faergeek replied:
Fixed, thanks

justinmk commented on cmake/RunTests.cmake:
```diff
@@ -64,6 +64,14 @@ endif()
+# In general $SHELL value can be pretty arbitrary and it seems like any
+# non-POSIX-compliant shell could cause some issues during a test run.
+#
+# Unsetting $SHELL makes neovim fall back to default value for 'shell' option.
```
I don't see why we would want that. We set SHELL=sh in /test/old/ so why not use the same approach here?

faergeek replied:
If we want to replicate what's being done in /test/old then we should set `$SHELL` to `sh` here and then set it to an empty string in `clear` only for Windows, right?

justinmk replied:
Let's keep the decision in one place (either RunTests or clear()). This should work:

```
if(NOT WIN32)
  set(ENV{SHELL} sh)
endif()
```

faergeek replied:
I updated this part. Seems to work just as well. I also changed comment a bit.

faergeek committed:
fix(tests): ignore $SHELL value in tests

Fixes #6172

"echo -n" was replaced by "printf", because "echo" in sh may be provided
as a shell builtin, which does not accept an "-n" flag to avoid a
trailing newline (e.g. on macos). "printf" is being used instead,
because it's more portable (defined by POSIX) and it does not output a
trailing newline by itself.

justinmk commented on cmake/RunTests.cmake:
```diff
@@ -64,6 +64,13 @@ endif()
 set(ENV{SYSTEM_NAME} ${CMAKE_HOST_SYSTEM_NAME})  # used by test/helpers.lua.
 
+if(NOT WIN32)
+  # Some tests may fail on fish shell #24941 #6172
+  # And in general $SHELL value can be pretty arbitrary and cause issues
+  # during a test run. Setting it to sh aims to avoid those kinds of issues.
```
can be much less verbose without losing information:

```suggestion
  # Tests assume POSIX "sh" and may fail if SHELL=fish. #24941 #6172
```

justinmk committed:
Update cmake/RunTests.cmake
